// @axa-fr/react-oidc-core/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var routes_1 = require("./routes");
exports.withRouter = routes_1.withRouter;
exports.OidcRoutes = routes_1.OidcRoutes;
var default_component_1 = require("./default-component");
exports.Authenticating = default_component_1.Authenticating;
exports.Callback = default_component_1.Callback;
var configurationPropTypes_1 = require("./configurationPropTypes");
exports.configurationPropTypes = configurationPropTypes_1.configurationPropTypes;
exports.configurationDefaultProps = configurationPropTypes_1.configurationDefaultProps;
var services_1 = require("./services");
exports.getUserManager = services_1.getUserManager;
exports.authenticationService = services_1.authenticationService;
exports.authenticateUser = services_1.authenticateUser;
exports.signinSilent = services_1.signinSilent;
exports.logoutUser = services_1.logoutUser;
exports.isRequireAuthentication = services_1.isRequireAuthentication;
exports.setLogger = services_1.setLogger;
exports.oidcLog = services_1.oidcLog;
exports.InMemoryWebStorage = services_1.InMemoryWebStorage;
